### PR TITLE
remove thread-locals in GenericIndexed in favor of wrapped objects

### PR DESCRIPTION
**Pull request by xvrl**
This should help reduce GC pressure form ThreadLocals on nodes with lots of indices. It may also slightly help with performance thanks to less branching.

### CODE REVIEW

File: processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java [hunk 1]
```diff
@@ -218,6 +218,8 @@ public void remove()
 
   private class CompressedIndexedLongs implements IndexedLongs
   {
+    final CloseableIndexed<ResourceHolder<LongBuffer>> cachedLongBuffers = baseLongBuffers.getCachingIndexed(0);
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java]: i found this a little confusing and had to look at the getCachingIndexed comments to understand what was happening. Is there a way to have another method for the special case when lookupCacheSize == 0?

File: processing/src/main/java/io/druid/segment/data/GenericIndexed.java [hunk 2]
```diff
@@ -327,23 +322,57 @@ public void writeToChannel(WritableByteChannel channel) throws IOException
...
-   * @param maxBytes maximum size in bytes of the lookup cache
-   * @return a copy of this GenericIndexed with a lookup cache.
+   * A caching indexed may perform better than the underlying Indexex even without a value cache
+   * since it does not have to guarantee thread-safety.
+   *
+   * @param lookupCacheSize maximum size in bytes of the lookup cache if greater than zero
+   * @return an Indexed wrapping the underlying indexed with a lookup cache.
    */
-  public GenericIndexed<T> withCache(int maxBytes)
+  public CloseableIndexed<T> getCachingIndexed(final int lookupCacheSize)
```
Comment (fjy) [inline processing/src/main/java/io/druid/segment/data/GenericIndexed.java]: will we ever know if a closeable indexed isn't closed somewhere or can another leak occur? I guess I'm wondering if there's a way to check if resources are being closed and throw an exception if they are not.

**Comment by cheddar** [comment]
Two things on this commit.

1) I'd like to request that we don't merge it until I've had a chance to merge in our current working branch.  I've already backed out the ThreadLocal changes there and am worried about losing important things in figuring out the merge conflicts

2) I think this patch takes a step closer to a less-invasive strategy for implementing the reuse of the buffer as well caching.  But, I think there's a cleaner implementation yet.  Specifically, I propose adding a method to the Indexed interface which is something like

```
public Indexed singleThreaded()
```

That essentially returns an object which is "optimized" given the knowledge that it is running off of a single thread.  That method call could essentially create a new Indexed object that wraps a reference to a single copy of the ByteBuffer which it reuses.

We can then move the caching behavior completely out of the Indexed and use composition.  I.e. make a `CachingIndexed` which wraps the object in an implementation that does caching.

This structure has added benefits in that the ColumnConfig no longer needs to be pushed down so far into the segment loading code and can instead be used at a much higher level.  Also, each part should be easier to unit test, I think.

Lastly, just a note, but if you are ever taking one interface, attaching the "Closeable" interface to it and making a new "CloseableXYZ" interface, you are going to eventually leak resources.  As a pattern, it should never be followed.  The reason is that implementations of `CloseableXYZ` are also implementations of `XYZ`, at some point they will be passed as just an `XYZ` instead of a `CloseableXYZ` losing the information that there is a `close()` method that needs to be dealt with.  If the interface needs a `close()` method, then ***all*** implementations of it need a close() method.

**Comment by xvrl** [comment]
@cheddar I believe the latest revision addresses your comments

File: processing/src/main/java/io/druid/segment/data/GenericIndexed.java [hunk 3]
```diff
@@ -179,133 +164,108 @@ public V getValue(Object key)
...
-    if (index == 0) {
-      startOffset = 4;
-      endOffset = copyBuffer.getInt(indexOffset);
-    } else {
-      copyBuffer.position(indexOffset + ((index - 1) * 4));
-      startOffset = copyBuffer.getInt() + 4;
-      endOffset = copyBuffer.getInt();
-    }
+      if (index == 0) {
+        startOffset = 4;
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/data/GenericIndexed.java]: Please make this a derivative number (`Integer.SIZE/8` or something)

File: processing/src/main/java/io/druid/segment/data/CachingIndexed.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+  private final GenericIndexed<T>.BufferIndexed indexed;
+  private final SizedLRUMap<Integer, T> cachedValues;
+  
+  public CachingIndexed(GenericIndexed<T>.BufferIndexed indexed, final int lookupCacheSize)
+  {
+    this.indexed = indexed;
+
+    if(lookupCacheSize > 0) {
+      log.debug("Allocating column cache of max size[%d]", lookupCacheSize);
+      cachedValues = new SizedLRUMap<>(INITIAL_CACHE_CAPACITY, lookupCacheSize);
```
Comment (drcrallen) [inline processing/src/main/java/io/druid/segment/data/CachingIndexed.java]: `SizedLRUMap` does not look thread safe. Upon cursory view of the construction below, it would seem like the thread safety were dependent on the dictionary, which is not the case. Can you either make this class thread safe, or mark it as NOT thread safe?

File: processing/src/main/java/io/druid/segment/data/GenericIndexed.java (hunk 3 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/data/GenericIndexed.java]: I would rather limit the changes to what is required to address the problem and not mix in unrelated changes, in order to make it clear this commit did not change the underlying logic.

File: processing/src/main/java/io/druid/segment/data/CachingIndexed.java (hunk 4 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/data/CachingIndexed.java]: Forgot to port the comments over from the old code

File: processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnSupplier.java [hunk 5]
```diff
@@ -52,7 +53,7 @@ public DictionaryEncodedColumn get()
     return new SimpleDictionaryEncodedColumn(
         singleValuedColumn,
         multiValuedColumn,
-        lookupCacheSize > 0 ? dictionary.withCache(lookupCacheSize) : dictionary
+        new CachingIndexed<>(dictionary, lookupCacheSize)
```
Comment (cheddar) [inline processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnSupplier.java]: Fwiw, keeping the original logic that has the if statement outside seems nicer to me.

File: processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java [hunk 6]
```diff
@@ -216,6 +216,8 @@ public void remove()
 
   private class CompressedIndexedLongs implements IndexedLongs
   {
+    final Indexed<ResourceHolder<LongBuffer>> cachedLongBuffers = baseLongBuffers.singleThreaded();
```
Comment (cheddar) [inline processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java]: This seems like a really weird name for this object given that it won't actually be cached by the object.  That's made even weirder by the fact that `CompressedIndexedLongs` is actually caching the result already.

File: processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java [hunk 7]
```diff
@@ -30,17 +30,17 @@
...
   private final VSizeIndexedInts column;
   private final VSizeIndexed multiValueColumn;
-  private final GenericIndexed<String> lookups;
+  private final CachingIndexed<String> cachedLookups;
 
   public SimpleDictionaryEncodedColumn(
       VSizeIndexedInts singleValueColumn,
       VSizeIndexed multiValueColumn,
-      GenericIndexed<String> lookups
+      CachingIndexed<String> cachedLookups
```
Comment (cheddar) [inline processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java]: I don't really agree with forcing all Dictionaries to be cached objects.  I really think that forcing the column to make the caching decision is incorrect.  It should be the query processing that makes the decision.  It is the segment/column's job to just provide information on what is there, ***not*** to optimize processing.

**Comment by cheddar** [comment]
While this PR is a step forward, two more things should be done:

1) Adding `singleThreaded` to the Indexed interface and moving the usage code out of the Column
2) The removal of `ColumnConfig` and all of the cruft it added to the Index loading code

Right now, this PR is mixing concerns by putting all of the caching into the columns

File: processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java (hunk 7 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/column/SimpleDictionaryEncodedColumn.java]: I agree, but I would rather fix that as part of a separate PR rather than changing the existing behavior

File: processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnSupplier.java (hunk 5 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnSupplier.java]: sounds good

**Comment by xvrl** [comment]
Regarding 1), the column cache currently relies on the size of values in the buffer as a shorthand to determine object sizes, so it is currently tied to GenericIndexed, and so it might not be straightforward to add `singleThreaded` to the Indexed interface.

I would prefer to do 2) as a separate PR to make the change in behavior clear.

File: processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnSupplier.java (hunk 5 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/serde/DictionaryEncodedColumnSupplier.java]: the indexed passed to the query has to be closable

File: processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java (hunk 6 above)
Comment (xvrl) [inline processing/src/main/java/io/druid/segment/data/CompressedLongsIndexedSupplier.java]: will fix that

**Comment by fjy** [comment]
+1 while I agree with @cheddar that some of this logic should be reworked, I think this PR makes progress on fixing an existing memory leak and we can work on some of @cheddar's comments in a future PR. I would want an issue filed for this work though, so we don't just forget about it.

**Comment by cheddar** [comment]
Hrm, ok, yeah, I can see your point.  I'm down with doing this for now and then following up in another PR to clean up various aspects.

**Comment by fjy** [comment]
Please remember to file an issue.
